### DISCUSSION TITLE
Examples of complex CustomTkinter applications.

### DISCUSSION THREAD

**Discussion by StuSerious** [TomSchimansky/CustomTkinter#1675] [Q&A]
Hey there!

I have recently begun work on [threedy](https://github.com/StuSerious/threedy), an application that I always wanted but could never quite find, made for interacting and working with text-based machine code files.

Keep in mind this is my first "ambitious" project where I want to actually make a useful and well maintained tool for the 3D Printing academic community, there's a lot I kinda know and kinda don't and I'm generally inexperienced with python since I'm coming from C# and Rust.

I am now at a point where I consider the UI part of it done, thanks largely through drawing inspiration (and code 😁) from @TomSchimansky's awesome examples, but I find myself stuck at the logic department.

Anyways, let me get to the point..
Are there any complex logic CTk apps out there that one could point to as examples of best practices/ good methodology when scaling CTk apps to a larger extent?

I saw and was amused at some of @Akascape's non-OOP code, but that's a bit alien to me, weird as that may sound, as I'm so used to OOP logic and classes.

Anyways, if anything comes to mind, that would be greatly appreciated! 
Thanks bunches in advance!

**Comment by abhaymakes** [comment]
Hello @StuSerious, there aren't much complex projects built with customtkinter as far as i have seen, and if there are chances are they're written without using OOP, i've created some [projects](https://github.com/TheFallen-Cat/PyTube-GUI-Downloader) using OOP though, you can take a look if you want.

However, i can give you some tips if your project is going to be large, like it contains multiple frames, switchable windows, and the windows need to communicate with each other, you'd be better off without OOP. OOP might make your code look organized but as the number of files(Classes) increase you might face management issues and circular dependencies. So, if you can avoid using OOP then do, don't eliminate them completely, use it only if you think the code will be better with them.

Also, if your classes needs to access data from other classes, make sure to create a medium for them, like creating another file to access them and then import that file in your code where you need to access the data.

Hope it helps!

**Reply by StuSerious** [reply]
Thanks for the insight! Having a look through your code, I got a couple ideas for the type of hierarchy I want to setup.

I was also finally able to avoid circular imports in some logic by creating a medium file as you suggested!

In the end, I think a mixed approach is the way to go. I will push my changes to the repo today or tomorrow and update here when I do so, if you want to take a peek at the architecture I'm coming up with.

Thanks again for the tips!
Cheers! 🎉

**Reply by avalon60** [reply]
> Hey there!
> 
> I have recently begun work on [threedy](https://github.com/StuSerious/threedy), an application that I always wanted but could never quite find, made for interacting and working with text-based machine code files.
> 
> Keep in mind this is my first "ambitious" project where I want to actually make a useful and well maintained tool for the 3D Printing academic community, there's a lot I kinda know and kinda don't and I'm generally inexperienced with python since I'm coming from C# and Rust.
> 
> I am now at a point where I consider the UI part of it done, thanks largely through drawing inspiration (and code grin) from @TomSchimansky's awesome examples, but I find myself stuck at the logic department.
> 
> Anyways, let me get to the point.. Are there any complex logic CTk apps out there that one could point to as examples of best practices/ good methodology when scaling CTk apps to a larger extent?
> 
> I saw and was amused at some of @Akascape's non-OOP code, but that's a bit alien to me, weird as that may sound, as I'm so used to OOP logic and classes.
> 
> Anyways, if anything comes to mind, that would be greatly appreciated! Thanks bunches in advance!

I'm not exactly sure what it is you are aiming for, but I'll take a stab. 

If you mean, you are looking at a way of organising your code with larger applications, I would strongly recommend you start looking at an MVC pattern (Module-View-Controller). With the exception of my latest app (CTk Theme Builder - there are reasons that I haven't done it  there yet), I have gone for this paradigm, because it makes your code far more manageable and so much easier to maintain, particularly where you have apps over a few thousand lines of code.

So:

1. Module :  responsible got controlling your data layer 
2. View: Used to control your data presentation (GUI)
3. Controller: This acts as the entry point and controls the other two.

I implement each of the above as a separate file, and each defines a class to do the job. You could just have it all in one file, but that kind of makes it less manageable.

I can't say I have gotten this perfect yet, but I am improving as I go on. The trick is to have the controller instantiate the Module and View classes. You also need to have the Controller package, pass itself to the View class as a parameter, on instantiation: 

```
import dccm_m as mod   # Import Module (data layer)
import dccm_v as vew     # Import View (GUI layer)

class DCCMView(ctk.CTk):
    """Class to instantiate our DCCM visual interface."""

    def __init__(**self**, **mvc_controller**, *args, **kwargs)
```
So as you see above, the View class expects to be  passed a reference to controller object.

From within the controller class:
            # Instantiate the GUI
            self.mvc_view = vew.DCCMView(mvc_controller=self)
            # Instantiate the data management layer
            self.mvc_module = mod.DCCMModule()

This allows the View module, access to the Controller methods, which in turn can grab data from the Module (data) layer. The aim is to **minimise** logic in the View module. You create methods as required in the View module, but these are orchestrated via the Controller class instance.

I have started a couple of projects and not used this approach, and ended up breaking them out and converting them. because the projects got big and unwieldy, so trust me, it's a good approach. 

Full code is [here](https://github.com/avalon60/dccm).

**Reply by StuSerious** [reply]
@avalon60 thank you Clive!

after reading your comment I've done a fair bit of reading regarding architecture patterns. I've decided to go for a MVP (Model-View-Presenter) since I think it fits best for the logic flow I have in mind.

[here is a short video going over some ezamples](https://www.youtube.com/watch?v=eHhXoCNCI1c)

Thanks again for your input!

**Reply by avalon60** [reply]
Cheers Stu. I have watched some of that guys videos before, he's very good.

**Comment by Akascape** [comment]
@StuSerious 
So, you need a complex application built with classes? 

Here is one program I made few months ago: https://github.com/Akascape/ShaderToyVid

**Reply by MaxNiark** [reply]
hello, 

i'm actually working on app. this is a big app for lot of thing. 

like you can check in the pict, i'm working on multiple file .py, and its workiing rly fucking good (auto py to exe work good too)

one thing is annoying=>  the auto scaling width and height, it's needed to create auto scaling function from .winfo_width/height, or C_get(*KWARG).

use .grid() method on the item in cls_frame to popup the frame, and grid_forget() for delete. dat all :)

![image](https://github.com/TomSchimansky/CustomTkinter/assets/126392328/278487a8-99fc-4aad-b523-19dd8cd665ed)
![image](https://github.com/TomSchimansky/CustomTkinter/assets/126392328/e7de6e8a-a37b-46b4-8ee3-1d8b5907ab27)
![image](https://github.com/TomSchimansky/CustomTkinter/assets/126392328/cdbb4cde-b52e-4a63-98d5-bfacc66ba284)
![image](https://github.com/TomSchimansky/CustomTkinter/assets/126392328/6580862c-5f8a-4446-b968-863230c50041)
